src/operator/numpy/linalg/broadcast_reduce_customized-inl.cuh

```diff
@@ -0,0 +1,416 @@
+          #pragma unroll
+          for (int u=0;u < unroll;u++) {
+            if (k + u*by < Mend) {
+              tmp[u] = OP1::Map(big[idx_big[u]], OP2::Map(lhs[idx_lhs[u]], rhs[idx_rhs[u]]));
+            }
+          }
+          #pragma unroll
+          for (int u=0;u < unroll;u++) {
+            if (k + u*by < Mend) reducer->Reduce(val, tmp[u], residual);
+          }
+        }
+      }
+
+      // Shared memory block bx * by. Reduction is along by. Final result is in tidy=0
+      if (by > 1) {
+        // Fix bx to avoid bank conflicts. Assumes warpSize number of banks
+        const int fbx = (do_transpose && ((bx & (warpSize - 1)) == 0)) ? (bx + 1) : bx;
+        const int it0 = tidx + tidy*fbx;
+        shTile[it0 * 2] = val;
+        shTile[it0 * 2 + 1] = residual;
+        __syncthreads();
+        for (int t=1;t < by;t <<= 1) {
+          DType tmp, tmp_residual;
+          reducer->SetInitValue(tmp, tmp_residual);
+          if (tidy + t < by) {
+            tmp = shTile[(it0 + t*fbx) * 2];
+            tmp_residual = shTile[(it0 + t*fbx) * 2 + 1];
+          }
+          __syncthreads();
+          reducer->Merge(shTile[it0 * 2], shTile[it0 * 2 + 1], tmp, tmp_residual);
+          __syncthreads();
+        }
+        if (idx < N && tidy == 0) {
+          reducer->Finalize(shTile[tidx * 2], shTile[tidx * 2 + 1]);
+          assign(&small[idx + m0*N], addto, shTile[tidx * 2]);
+        }
+      } else {
+        if (idx < N) {
+          reducer->Finalize(val, residual);
+          assign(&small[idx + m0*N], addto, val);
+        }
+      }
+    }
+  }
+  if (need_clean) {
+    delete reducer;
+  }
+}
+
+// Simple reduction of lines when M is small
+template<typename Reducer, typename DType>
+__launch_bounds__(kMaxThreadsPerBlock)
+__global__ void reduce_lines_kernel_wr(const int N, const int M, const bool addto,
+  const int small_in_stride, const DType* __restrict small_in, DType *small_out,
+  Reducer* reducer = nullptr) {
+  bool need_clean = !reducer;
+  reducer = reducer ? reducer : new Reducer();
+  for (int idx = threadIdx.x + blockIdx.x*blockDim.x; idx < N; idx += blockDim.x*gridDim.x) {
+
+    DType val, residual;
+    reducer->SetInitValue(val, residual);
+    for (int k = 0; k < M; k++) {
+      reducer->Reduce(val, small_in[idx + k*small_in_stride], residual);
+    }
+
+    if (idx < N) {
+      reducer->Finalize(val, residual);
+      assign(&small_out[idx], addto, val);
+    }
+
+  }
+  if (need_clean) {
+    delete reducer;
+  }
+}
+
+template<typename Reducer, int ndim, typename AType, typename DType, typename OType, typename OP>
+__launch_bounds__(kMaxThreadsPerBlock)
+__global__ void reduce_kernel_M1_wr(const int N, const bool addto,
+                                    const DType* __restrict big, OType *small, const Shape<ndim> bshape,
+                                    const Shape<ndim> sshape, Reducer* reducer = nullptr) {
+  bool need_clean = !reducer;
+  reducer = reducer ? reducer : new Reducer();
```

Comment:
I think we can probably move all those reducer creations out of the kernels and put on the host side.

Reply:
Absolutely.